Configurable trailing slash behaviour for NormalizePath

l-2-j opened the pull request:
<!-- Thanks for considering contributing actix! -->
<!-- Please fill out the following to make our reviews easy. -->

## PR Type
<!-- What kind of change does this PR make? -->
<!-- Bug Fix / Feature / Refactor / Code Style / Other -->
Feature


## PR Checklist
Check your PR fulfills the following:

<!-- For draft PRs check the boxes as you complete them. -->

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt


## Overview
<!-- Describe the current and new behavior. -->
Currently NormalizePath always adds a trailing slash. This change would make it configurable to either follow that style, or to always trim the trailing slash instead.
<!-- Emphasize any breaking changes. -->
Will be a breaking change to `NormalizePath` usage, since `middleware::NormalizePath` will need to be replaced by `middleware::NormalizePath::default()` or `::new(...)` when creating the middleware.


<!-- If this PR fixes or closes an issue, reference it here. -->
Closes #1612

Commit:
Configurable behaviour for NormalizePath

robjtede commented on src/middleware/normalize.rs:
```diff
@@ -50,13 +72,18 @@ where
 pub struct NormalizePathNormalization<S> {
     service: S,
     merge_slash: Regex,
+    trim_last_slash: bool,
```
Given how we're using it, makes sense to keep the enum here too. Makes for more expressive matches later and possible ease of extensibility.

robjtede commented on src/middleware/normalize.rs:
```diff
@@ -32,7 +41,20 @@ use crate::Error;
+    fn default() -> Self {
+        NormalizePath::new(TrailingSlash::Always)
+    }
+}
```
If we impl Default for TrailingSlash we can keep the derive on NormalizePath.

robjtede commented on src/middleware/normalize.rs:
```diff
@@ -10,20 +10,29 @@ use regex::Regex;
+/// To be used when constructing `NormalizePath` to define it's behavior.
+#[derive(Clone, Copy)]
+pub enum TrailingSlash {
+    /// Always add a trailing slash to the end of the path.
```
Make a note that all routes defined must end in trailing slash for them to to be accessible.

robjtede commented on src/middleware/normalize.rs:
```diff
@@ -150,14 +180,36 @@ mod tests {
     }
 
+    #[actix_rt::test]
+    async fn trim_trailing_slashes() {
```
Also include a test request without trailing slash.

robjtede commented on src/middleware/normalize.rs:
```diff
@@ -78,8 +105,11 @@ where
+        // Either adds a string to the end (duplicates will be removed anyways) or trims all slashes from the end
+        let path = match self.trim_last_slash {
+            false => original_path.to_string() + "/",
+            true => original_path.trim_end_matches(|c| c == '/').to_string(),
```
I believe this is equivalent.

```suggestion
            true => original_path.trim_end_matches('/').to_string(),
```

robjtede commented on src/middleware/normalize.rs:
```diff
@@ -10,20 +10,29 @@ use regex::Regex;
-#[derive(Default, Clone, Copy)]
+/// To be used when constructing `NormalizePath` to define it's behavior.
+#[derive(Clone, Copy)]
+pub enum TrailingSlash {
```
Can you stick a #[non_exhaustive] on this please. Unlikely folks will want to match on it but would let us do things in future without breaking changes.

robjtede reviewed (changes requested):
Looks really good so far. Just left some initial thoughts and obviously the checks are failing from the doc comment in condition.rs.

Commit:
Changes & fixes based on feedback

Made TrailingSlash be non_exhaustive with a default of Always, and changed NormalizePath back to derive. Started storing the TrailingSlash enum in NormalizePathNormalization instead of a boolean. Added one more test request without a trailing slash. And also tidied up a bit.

l-2-j committed:
Merge branch 'master' into trailing-slash-normalization

robjtede commented:
Thanks for those updates. Add a CHANGELOG.md entry and a note to MIGRATION.md. Also check the job failure to see what needs updating.

l-2-j committed:
Fixed NormalizePath usage in docs

l-2-j committed:
Added details to markdown files about NormalizePath changes

robjtede commented:
Great work, thanks 👍

patrickelectric commented:
Hey, any plan to release it ? It's an awesome feature, I was able to have a `/path|path/` to the same get method, and now with 2.0.0 appears to be not possible at all sadly.
